File: Backend/app/schemas/analysis.py
"""Analysis Pydantic schemas."""

from __future__ import annotations

import uuid
from datetime import datetime
from typing import Any

from pydantic import BaseModel, Field

from app.core.constants import AnalysisStatus


class AnalysisRequest(BaseModel):
    resume_id: uuid.UUID
    job_id: uuid.UUID


class ScoreBreakdown(BaseModel):
    keyword: float | None = None
    semantic: float | None = None
    skill_depth: float | None = None
    experience: float | None = None
    impact: float | None = None
    education: float | None = None


class ConfidenceInterval(BaseModel):
    lower: float
    upper: float
    confidence: float


class KeywordDetail(BaseModel):
    matched_required: list[str] = Field(default_factory=list)
    missing_required: list[str] = Field(default_factory=list)
    matched_preferred: list[str] = Field(default_factory=list)
    match_rate: float = 0.0


class SemanticDetail(BaseModel):
    per_requirement_scores: dict[str, float] = Field(default_factory=dict)
    mean_score: float = 0.0
    strong_matches: list[str] = Field(default_factory=list)
    weak_matches: list[str] = Field(default_factory=list)


class ExperienceDetail(BaseModel):
    total_yoe: float | None = None
    required_yoe: float | None = None
    seniority_inferred: str | None = None
    seniority_required: str | None = None
    career_progression_score: float | None = None


class AnalysisRead(BaseModel):
    id: uuid.UUID
    resume_id: uuid.UUID
    job_id: uuid.UUID
    status: AnalysisStatus
    score: float | None = Field(default=None, alias="score", serialization_alias="score_composite")
    confidence: float | None = None
    percentile: float | None = None
    created_at: datetime
    completed_at: datetime | None = None
    celery_task_id: str | None = None

    model_config = {"from_attributes": True, "populate_by_name": True}


class AnalysisResultRead(AnalysisRead):
    score_breakdown: ScoreBreakdown | None = None
    confidence_interval: ConfidenceInterval | None = None
    keyword_detail: KeywordDetail | None = None
    semantic_detail: SemanticDetail | None = None
    experience_detail: ExperienceDetail | None = None
    ats_warnings: list[str] | None = Field(default_factory=list)
    pipeline_meta: dict[str, Any] | None = None


class AnalysisCreateResponse(BaseModel):
    analysis_id: uuid.UUID
    status: AnalysisStatus
    message: str = "Analysis queued. Poll /analysis/{id} for status."

File: Frontend/src/types/analysis.ts
export type AnalysisStatus = 'QUEUED' | 'PARSING' | 'MATCHING' | 'SCORING' | 'FEEDBACK' | 'DONE' | 'FAILED' | 'PARTIAL'
export type FeedbackSeverity = 'critical' | 'high' | 'medium' | 'low'
export type FeedbackCategory = 'keyword' | 'semantic' | 'impact' | 'ats' | 'education' | 'experience' | 'formatting'

export interface AnalysisRequest {
  resume_id: string
  job_id: string
}

export interface AnalysisCreateResponse {
  analysis_id: string
  status: AnalysisStatus
  message: string
}

export interface AnalysisRead {
  id: string
  resume_id: string
  job_id: string
  status: AnalysisStatus
  score_composite: number | null
  confidence: number | null
  percentile: number | null
  created_at: string
  completed_at: string | null
  celery_task_id: string | null
}

export interface ScoreBreakdown {
  keyword?: number
  semantic?: number
  skill_depth?: number
  experience?: number
  impact?: number
  education?: number
}

export interface AnalysisResultRead extends AnalysisRead {
  score_breakdown: ScoreBreakdown | null
  confidence_interval?: {
    lower: number
    upper: number
    confidence: number
  }
  keyword_detail?: {
    matched_required: string[]
    missing_required: string[]
    matched_preferred: string[]
    match_rate: number
  }
  semantic_detail?: {
    per_requirement_scores: Record<string, number>
    mean_score: number
    strong_matches: string[]
    weak_matches: string[]
  }
  experience_detail?: {
    total_yoe?: number
    required_yoe?: number
    seniority_inferred?: string
    seniority_required?: string
    career_progression_score?: number
  }
  ats_warnings: string[]
  pipeline_meta: Record<string, unknown> | null
}

export interface FeedbackItemRead {
  id: string
  analysis_id: string
  category: FeedbackCategory
  severity: FeedbackSeverity
  title: string
  description: string
  original_text: string | null
  suggested_text: string | null
  score_delta: number | null
  source_section: string | null
  accepted: boolean | null
  created_at: string
}

export interface RewriteRequest {
  feedback_item_id: string
  section: string
}

export interface RewriteResult {
  feedback_item_id: string
  original_text: string
  rewritten_text: string
  hallucination_check_passed: boolean
  flagged_entities: string[]
  warning?: string
}
